Src/VimCore/CommonOperations.fs

```diff
@@ -446,8 +446,11 @@ type internal CommonOperations
 
         /// Move the caret right.  Don't go off the end of the line
         let moveRight () =
-            if x.CaretPoint.Position < x.CaretLine.End.Position then
-                let point = SnapshotPointUtil.AddOne x.CaretPoint
+            let point =
+                x.CaretPoint
+                |> SnapshotPointUtil.AddOne
+                |> (fun y -> if Char.IsSurrogate (y.GetChar()) then SnapshotPointUtil.AddOne y else y)
+            if point.Position <= x.CaretLine.End.Position then

```

Comment:
Think we need to add a similar fix for `CharLeft`.  Even though the caret functionally moves left today it's broken for other scenarios.  Consider when my caret is to the right of a surrogate pair.  In that case I would expect the normal mode command `yh` would yank the full unicode character.  Just as `yl` would yank the unicode character when the caret was at the start of the character. 

In order to do that `CharLeft` needs to fully encompass the character.